libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -107,10 +107,51 @@ size_t IRGenerationContext::immutableMemoryOffset(VariableDeclaration const& _va
 	return m_immutableVariables.at(&_variable);
 }
 
+size_t IRGenerationContext::immutableMemoryOffsetRelative(VariableDeclaration const& _variable) const
+{
+	auto const absoluteOffset = immutableMemoryOffset(_variable);
+	solAssert(absoluteOffset >= CompilerUtils::generalPurposeMemoryStart);
+	return absoluteOffset - CompilerUtils::generalPurposeMemoryStart;
+}
+
+size_t IRGenerationContext::reservedMemorySize() const
+{
+	solAssert(m_reservedMemory.has_value());
+	return *m_reservedMemory;
+}
+
+void IRGenerationContext::registerLibraryAddressImmutable()
+{
+	solAssert(m_reservedMemory.has_value(), "Reserved memory has already been reset.");
+	m_libraryAddressImmutableOffset = CompilerUtils::generalPurposeMemoryStart + *m_reservedMemory;
+	*m_reservedMemory += 32;
+}
+
+size_t IRGenerationContext::libraryAddressImmutableOffset() const
+{
+	solAssert(m_libraryAddressImmutableOffset.has_value());
+	return *m_libraryAddressImmutableOffset;
+}
+
+size_t IRGenerationContext::libraryAddressImmutableOffsetRelative() const
+{
+	solAssert(m_libraryAddressImmutableOffset.has_value());
+	solAssert(m_libraryAddressImmutableOffset >= CompilerUtils::generalPurposeMemoryStart);
+	return *m_libraryAddressImmutableOffset - CompilerUtils::generalPurposeMemoryStart;
+}
+
 size_t IRGenerationContext::reservedMemory()
 {
 	solAssert(m_reservedMemory.has_value(), "Reserved memory was used before.");
 	size_t reservedMemory = *m_reservedMemory;
+
+	// We assume reserved memory contains only immutable variables.
+	// This memory is used i.e. by RETURNCONTRACT to create new EOF container with aux data.
+	size_t immutableVariablesSize = 0;
+	for (auto const* var: keys(m_immutableVariables))
+		immutableVariablesSize += var->annotation().type->memoryHeadSize();
```

Comment:
`memoryHeadSize()` is not what we want. For reference types it will return the size of the pointer rather than the whole type. We have `memoryDataSize()`, but that's probably not correct either, because for reference types nested inside an array/struct we'll still get just the size of the pointer included. The actual size will actually depend on the layout we'll choose, which may actually differ from what we use for storage, memory or calldata.

So my original suggestion to calculate total size here is actually problematic. Let's not do it after all and make it work only for value types. Better to leave calculating it correctly for the time when we'll actually know the layout.

```suggestion
		{
			solUnimplementedAssert(var->type()->isValueType());
			solUnimplementedAssert(var->type()->sizeOnStack() == 1);		
			immutableVariablesSize += var->type()->sizeOnStack() * 32;
		}
```

Reply:
done